### PR TITLE
New SCC scheduler

### PR DESCRIPTION
**Pull request by amir734jj**
- SCC chunk scheduler
- Static code generation using the SCC scheduler + `do { } while();` loops
- Conditional comments in the generated scala code
- No DOWN-UP in the aps-cycle, only UP-DOWN if there is only a fiber cycle
- Detect unconditional cycles and throw an error (not implemented yet)
- Does not break any existing APS code flow

To use static SCC chunk scheduling:

```shell
../bin/apssched -S -DCOT -p .:../base nested-cycles      
../bin/aps2scala -SCC -DCOT -p .:../base nested-cycles
```

### CODE REVIEW

File: analyze/aps-schedule.c [hunk 1]
```diff
@@ -0,0 +1,2989 @@
...
+  if (current == NULL)
+    return;
+
+  CHILD_PHASE* current_group = &current->child_phase;
+
+  if (IS_VISIT_MARKER(current)) {
+    if (current->child_phase.ch == -1) {
+      // End of total order
+      if (current->cto_next == NULL)
+        return;
```
Comment (boyland) [inline analyze/aps-schedule.c:282]: I don't understand.  I would think that current->cto_next should be null unless this is the parent INHERITED phase which I think doesn't use a marker.

So why not have an error message if cto_next is *not* null.

Also: after this you have a number of checks.  Those only happen if the cto_next is not null.  Do you not need to check them if cto_next == null?

File: analyze/aps-schedule.c [hunk 2]
```diff
@@ -0,0 +1,2989 @@
...
+      followed_by(current->cto_next, current->child_phase.ph,
+                  current->child_phase.ch, true /* immediate */,
+                  false /* not visit markers only*/,
+                  &followed_by_child_synthesized);
+
+      // Boolean indicating whether visit marker was followed by child
+      // inherited attribute(s) <-ph,ch>
+      bool preceded_by_child_inherited = prev_group->ph == -current_group->ph &&
+                                         prev_group->ch == current_group->ch;
+
```
Comment (boyland) [inline analyze/aps-schedule.c:330]: Not sure how you handle the case when the visit has no attributes at all or only syntax or only int

File: aps2scala/dump-scala.cc [hunk 3]
```diff
@@ -92,6 +94,35 @@ static string program_id(string name)
   return result;
 }
 
+/**
+ * Utility function to dump static circular evaluation trait to
+ * by pass some of CircularEvaluation limitations
```
Comment (boyland) [inline aps2scala/dump-scala.cc:99]: What's wrong with CircularEvaluation?  Maybe it should be fixed, not just extended?

**Review by boyland — COMMENTED** [review]
Some comments

**Comment by boyland** [comment]
Maybe use -C overall for your circular scheduler.  I don't think it's being used.  "-SCC" looks like three options.  And sometimes "-S" and sometimes "-SCC" is confusing.

File: analyze/aps-analyze.c [hunk 4]
```diff
@@ -38,27 +43,55 @@ static void *analyze_thing(void *ignore, void *node)
...
+        // Pure fiber cycles should have been broken when reaching this step
+        // If there is a non-monotone cycle, then scheduling is no longer possible
+        if (d & DEPENDENCY_MAYBE_SIMPLE)
+        {
+          if (cycle_debug & PRINT_CYCLE)
+          {
+            print_cycles(s, stdout);
+          }
+
+          aps_error(decl, "Non-monotone Cycle detected (%d); Attribute grammar is not OAG", d);
```
[the remark this answers is not in the document]
Comment (amir734jj) [reply analyze/aps-analyze.c]: Fixed the typo. Sorry

File: analyze/aps-cycle.c [hunk 5]
```diff
@@ -900,10 +900,28 @@ void break_fiber_cycles(Declaration module,STATE *s,DEPENDENCY dep) {
...
+    else
+    {
+      add_up_down_attributes(s,UP_DOWN);
+    }
+  }
+  else
+  {
+    // Preserve UP-DOWN edges if the accumulated dependency is just fiber cycle,
+    // otherwise preserve DOWN-UP edges.
+    bool direction = !(dep & DEPENDENCY_NOT_JUST_FIBER) ? UP_DOWN : DOWN_UP;
```
[the remark this answers is not in the document]
Comment (amir734jj) [reply analyze/aps-cycle.c:931]: Added a paragraph TODO describing the situation, why we don't need `DOWN-UP` anymore, what was the problem with `DOWN-UP`, and how we tried adding code to ensure `fiber-cycle-breaking` doesn't remove direct edges.

File: analyze/aps-dnc.h [hunk 6]
```diff
@@ -1,6 +1,7 @@
 #ifndef APS_DNC_H
 #define APS_DNC_H
 
+#include <stdbool.h>
```
[the remark this answers is not in the document]
Comment (amir734jj) [reply analyze/aps-dnc.h]: oops, sorry for the typo

File: analyze/aps-schedule.c [hunk 7]
```diff
@@ -0,0 +1,2989 @@
...
+    if ((visit_marker_only && IS_VISIT_MARKER(current)) || !visit_marker_only) {
+      *any = true;
+    }
+  } else if (immediate) {
+    return;
+  }
+
+  bool if_true_any = false;
+  bool if_false_any = false;
+
```
[the remark this answers is not in the document]
Comment (amir734jj) [reply analyze/aps-schedule.c]: Fixed it and used the same `any` variable. It was unnecessary.

File: analyze/aps-schedule.c [hunk 8]
```diff
@@ -0,0 +1,2989 @@
...
+ * @param ch child number
+ * @return boolean indicating if there is any
+ */
+static bool aug_graph_contains_phase(AUG_GRAPH* aug_graph,
+                                     TOTAL_ORDER_STATE* state,
+                                     CONDITION cond,
+                                     const short ph,
+                                     const short ch) {
+  int i;
+  bool exist_parent_inherited_attr_previous_phase = false;
```
[the remark this answers is not in the document]
Comment (amir734jj) [reply analyze/aps-schedule.c]: It was not used. So sorry.

File: aps2scala/aps2scala.cc [hunk 9]
```diff
@@ -56,11 +57,16 @@ int main(int argc,char **argv) {
     } else if (streq(argv[i],"-S") || streq(argv[i],"--static")) {
       static_schedule = true;
       continue;
-    } else if (streq(argv[i],"-V") || streq(argv[i],"--verbose")) {
+    } else if (streq(argv[i],"-SCC") || streq(argv[i],"--static-scc")) {
+      static_schedule = true;
+      static_scc_schedule = true;
+      continue;
+    }  else if (streq(argv[i],"-V") || streq(argv[i],"--verbose")) {
```
[the remark this answers is not in the document]
Comment (amir734jj) [reply aps2scala/aps2scala.cc]: oops, so sorry. It's fixed now.

File: apscpp/apscpp.cc [hunk 10]
```diff
@@ -27,6 +27,7 @@ void usage() {
   fprintf(stderr,"    -DH   list debugging flags\n");
   fprintf(stderr,"    -V    increase verbosity of generation code\n");
   fprintf(stderr,"    -G    add Debug calls for every function\n");
+  fprintf(stderr,"    -S    SCC chunk static scheduling\n");
```
[the remark this answers is not in the document]
Comment (amir734jj) [reply apscpp/apscpp.cc]: Changes it to `-C` flag as you recommended.

File: aps2scala/dump-scala.cc (hunk 3 above)
Comment (amir734jj) [reply aps2scala/dump-scala.cc]: Added a description of _what would go wrong with_ `CircularEvaluation` **IF** we use it for static circular evaluation.

More specifically, `CircularEvaluation` doesn't record a `change` in attribute instance value and does not like attribute value being re-assigned.

File: analyze/aps-schedule.c (hunk 2 above)
Comment (amir734jj) [reply analyze/aps-schedule.c:326]: It's already handled I think. Local attributes don't affect `prev_group` in the recursive call and we check if there is indeed an attribute instance of `<-ph,ch>` in the instances of augmented dependency graph **and then** we throw an exception. So its completely okay for child visit to be empty OR not have inherited or synthesized attributes OR both.

File: analyze/aps-schedule.c [hunk 11]
```diff
@@ -0,0 +1,2989 @@
...
+}
+
+/**
+ * Utility function used by assert_total_order to check sanity of total order
+ *  1) After visit marker <ph,-1> there should be no attribute
+ *     belonging to <ph,-1> or <-ph,-1>
+ *  2) Immediately before visit marker <ph,ch> should be attribute
+ *     belonging to <-ph,ch>
+ *  3) Or immediately after visit marker <ph,ch> should be attribute belonging
+ *     to <ph,ch>
```
[the remark this answers is not in the document]
Comment (amir734jj) [reply analyze/aps-schedule.c]: Updated the comment. Scheduler and assertion both support empty visits.

Empty visits are the ones that don't have child-inherited or child-synthesized attributes before/after visit call and the assertion code doesn't blindly throw an exception if they are missing. It checks if such attribute instances belonging to these group exists before throwing an exception.

File: analyze/aps-schedule.c (hunk 1 above)
Comment (amir734jj) [reply analyze/aps-schedule.c]: You are correct. Updated the ordering. It prematurely exited the function before making sure before the parent end-of-phase visit marker there were parent synthesized attributes (if any).

This function checks the sanity of CTO nodes. More specifically that code checks if we are currently at end of the parent phase visit marker `<ph,-1>`

- This means there should be a parent synthesized attribute (if any) before the visit marker
- Parent-inherited attributes of the next phase (if any) should be followed by the visit marker. **However**, if this is the last phase then we should check if its followed by parent-inherited attributes of the next phase.


Note that the final CTO node is always the end of the parent visit marker of the last phase of the parent `<max(ph), -1>`

**Review by boyland — APPROVED** [review]
Looks good.  Ready to merge in?

**Pull request merged**